cpp/src/arrow/builder.cc

```diff
@@ -1149,15 +1181,20 @@ Status DictionaryBuilder<T>::Append(const Scalar& value) {
 
   RETURN_NOT_OK(values_builder_.Append(index));
 
+  is_finished_ = false;
   return Status::OK();
 }
 
 template <typename T>
 Status DictionaryBuilder<T>::AppendNull() {
+  is_finished_ = false;
   return values_builder_.AppendNull();
 }
 
-Status DictionaryBuilder<NullType>::AppendNull() { return values_builder_.AppendNull(); }
+Status DictionaryBuilder<NullType>::AppendNull() {
+  is_finished_ = false;
```

Comment:
Curious why this is needed

Reply:
There are nulls appended to the `values_builder_`, so it could resized and contain uninitialized padding, if we never call finish on it.